interfaces/apparmor/backend_test.go

```diff
@@ -0,0 +1,509 @@
+	const before = `
+name: samba
+version: 1
+developer: acme
+apps:
+    smbd:
+    nmbd:
+`
+	// NOTE: the version is the same so that no unrelated changes are made
+	const after = `
+name: samba
+version: 1
+developer: acme
+apps:
+    smbd:
+`
+	for _, developerMode := range []bool{true, false} {
+		snapInfo := s.installSnap(c, developerMode, before)
+		s.cmds["apparmor_parser"].ForgetCalls()
+		snapInfo = s.updateSnap(c, snapInfo, developerMode, after)
+		profile := filepath.Join(s.backend.Directory(), "snap.samba.nmbd")
+		// file called "snap.sambda.nmbd" was removed
+		_, err := os.Stat(profile)
+		c.Check(os.IsNotExist(err), Equals, true)
+		// apparmor_parser was used to remove the unused profile
+		c.Check(s.cmds["apparmor_parser"].Calls(), DeepEquals, []string{"--remove snap.samba.nmbd"})
+	}
+}
+
+// Low level tests for constituent parts of backend API
+
+// Tests for Backend.CombineSnippets()
+
+func (s *backendSuite) TestRealDefaultTemplateIsNormallyUsed(c *C) {
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(`
+name: SNAP
+version: VERSION
+developer: DEVELOPER
+apps:
+    APP:
+`))
+	c.Assert(err, IsNil)
+	// NOTE: we don't call apparmor.MockTemplate()
+	content, err := s.backend.CombineSnippets(snapInfo, false, nil)
+	c.Assert(err, IsNil)
+	profile := string(content["snap.SNAP.APP"].Content)
+	for _, line := range []string{
+		// NOTE: a few randomly picked lines from the real profile.  Comments
+		// and empty lines are avoided as those can be discarded in the future.
+		"#include <tunables/global>\n",
+		"/tmp/   r,\n",
+		"/sys/class/ r,\n",
+	} {
+		c.Assert(profile, testutil.Contains, line)
+	}
+}
+
+func (s *backendSuite) TestCustomTemplateUsedOnRequest(c *C) {
+	s.backend.CustomTemplate = `
+# Description: Custom template for testing
+###VAR###
+
+###PROFILEATTACH### (attach_disconnected) {
+	FOO
+}
+`
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(`
+name: SNAP
+version: VERSION
+developer: DEVELOPER
+apps:
+    APP:
+`))
+	c.Assert(err, IsNil)
+	content, err := s.backend.CombineSnippets(snapInfo, false, nil)
+	c.Assert(err, IsNil)
+	profile := string(content["snap.SNAP.APP"].Content)
+	// Our custom template was used
+	c.Assert(profile, testutil.Contains, "FOO")
+	// Custom profile can rely on legacy variables
+	for _, legacyVarName := range []string{
+		"APP_APPNAME", "APP_ID_DBUS", "APP_PKGNAME_DBUS",
+		"APP_PKGNAME", "APP_VERSION", "INSTALL_DIR",
+	} {
+		c.Assert(profile, testutil.Contains, fmt.Sprintf("@{%s}=", legacyVarName))
+	}
+}
+
+func (s *backendSuite) TestCombineSnippets(c *C) {
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(`
+name: SNAP
+version: VERSION
+developer: DEVELOPER
+apps:
+    APP:
+`))
+	c.Assert(err, IsNil)
+	glob := s.backend.FileGlob(snapInfo)
+	// NOTE: replace the real template with a shorter variant
+	restore := apparmor.MockTemplate("\n" +
+		"###VAR###\n" +
+		"###PROFILEATTACH### (attach_disconnected) {\n" +
+		"}\n")
+	defer restore()
+	for _, scenario := range []struct {
```

Comment:
I did my best to simplify this but I kept the strings as there are sadly each unique. Have a look if I can make anything else better.